editor/project_manager.cpp

```diff
@@ -1174,13 +1182,57 @@ void ProjectList::load_project_data(const String &p_property_key, Item &p_item,
 	}
 
 	if (config_version > ProjectSettings::CONFIG_VERSION) {
-		// Comes from an incompatible (more recent) Godot version, grey it out
+		// Comes from an incompatible (more recent) Godot version, gray it out.
 		grayed = true;
 	}
 
-	String description = cf->get_value("application", "config/description", "");
-	String icon = cf->get_value("application", "config/icon", "");
-	String main_scene = cf->get_value("application", "run/main_scene", "");
+	const String description = cf->get_value("application", "config/description", "");
+	const String icon = cf->get_value("application", "config/icon", "");
+	const String main_scene = cf->get_value("application", "run/main_scene", "");
+
+	PackedStringArray project_features = cf->get_value("application", "config/features", PackedStringArray());
+	bool project_features_dirty = false;
+	// If there is no feature list currently present, force one to generate.
+	if (project_features.is_empty()) {
+		project_features = ProjectSettings::get_required_features();
+		project_features_dirty = true;
+	}
+	// Check the rendering API.
+	const String rendering_api = cf->get_value("rendering", "quality/driver/driver_name", "Unknown");
+	if (rendering_api != "Unknown") {
+		// Add the rendering API as a project feature if it doesn't already exist.
+		if (!project_features.has(rendering_api)) {
+			project_features.append(rendering_api);
+			project_features.sort();
+			project_features_dirty = true;
+		}
+	}
+	// Check for the existence of a csproj file.
+	if (FileAccess::exists(path.plus_file(project_name + ".csproj"))) {
+		// If there is a csproj file, add the C# feature if it doesn't already exist.
+		if (!project_features.has("C#")) {
+			project_features.append("C#");
+			project_features.sort();
+			project_features_dirty = true;
+		}
+	} else {
+		// If there isn't a csproj file, remove the C# feature if it exists.
+		if (project_features.has("C#")) {
+			project_features.remove(project_features.find("C#"));
+			project_features.sort();
+			project_features_dirty = true;
+		}
+	}
+	// Write the updated feature list, but only if the project config version is the same as ours.
+	// We should never write to project files with a different config version.
+	if (project_features_dirty && config_version == ProjectSettings::CONFIG_VERSION) {
+		ProjectSettings *ps = memnew(ProjectSettings);
+		ps->load_custom(conf);
+		ps->set("application/config/features", project_features);
+		ps->save_custom(conf);
+		memdelete(ps);
+	}
```

Comment:
I noticed that you do `sort()` and set `project_features_dirty` to true each time. The sorting could just be done in this block, single time. Or in another condition block that only checks `project_features_dirty`.